File: blackJackGambler.py
import random as rand
import math

# This class has the following methods: fundChange, drawCard, cardSum, busted, and newRound
class blackJackGambler:
    def __init__(self, amtFunds):
        self.amtFunds = amtFunds
        self.bet = 0
        self.cards = []
        self.cardNames = []
        self.handTotal = []
        self.bust = False
        self.blackJack = False
        self.deck = list(range(0,51))

    def fundChange(self, outcome):  # When a round is over, the gamblers pot will either increase of decrease.
        if outcome==1:
            self.amtFunds+=self.bet
        else:
            self.amtFunds-=self.bet

    def drawCard(self): # A number from 0 to 51 is chosen at random, each number representing a card. This number (card) is then taken out of circulation until a new deck is needed
        if len(self.deck) > 0:
            randCard = rand.randint(0,len(self.deck)-1)
            self.cards.append(self.deck.pop(randCard))
        else:
            self.deck = list(range(0, 51))
            print("New Deck!")
            randCard = rand.randint(0,len(self.deck)-1)
            self.cards.append(self.deck.pop(randCard))

        self.cardSum()
    
    def cardSum(self):  # Takes the random integers in a gamblers cards and finds their associated real life card and value! It is a list in the event of a drawn ACE which can have two different value scenarios.
        sumCards = [0, 0]
        self.cardNames.clear()
        for c in self.cards:
            head = ""
            suit = math.floor(((c+1)/13))
            val = (c+1) - (suit*13)

            if suit <= 1:
                suit = "Hearts"
            elif suit <= 2:
                suit = "Spades"
            elif suit <= 3:
                suit = "Clubs"
            else:
                suit = "Diamonds"
            
            if val == 1:
                head = "ACE"
            elif val == 11:
                head = "KING"
                val = 10
            elif val == 12:
                head = "QUEEN"
                val = 10
            elif val == 0:
                head = "JACK"
                val = 10
            
            if head != "":
                cardName = f'{head} of {suit}'
            else:
                cardName = f'{val} of {suit}'
            self.cardNames.append(cardName)
            
            if head == "ACE":
                sumCards[0]+=1
                sumCards[1]+=11
            else:
                sumCards[0]+=val
                sumCards[1]+=val

        self.handTotal = sumCards
        self.busted()

    def busted(self):   # Determines if the gambler has a hand that is over 21 (in both scenarios in the event of an ACE) or not.
        if (self.handTotal[0] == 21) or (self.handTotal[1] == 21):
            self.blackJack = True
        else:
            if (self.handTotal[0] > 21) and (self.handTotal[1] > 21):
                self.bust = True

    def newRound(self):     # Initiates a new hand by taking away gamblers cards and values to set them up fresh for a new round.
        self.cards = []
        self.cardNames = []
        self.handTotal = []
        self.bust = False
        self.blackJack = False

File: blackJack/blackJack/blackJackGambler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;

namespace bj
{
    public class blackJackGambler
    {
        public float amtFunds = 0;
        public float bet = 0;
        public List<int> cards = new List<int>();
        public List<string> cardNames = new List<string>();
        public List<int> handTotal = new List<int>();
        public bool bust = false;
        public bool blackJack = false;
        public List<int> deck = new List<int>();


        public void deckCreation()
        {
            deck = Enumerable.Range(0, 52).ToList();
        }

        public void fundChange(int outcome)
        {
            if (outcome == 1) {
                amtFunds += bet;
            }
            else {
                amtFunds -= bet;
            }
        }

        public void drawCard()
        {
            Random random = new Random();
            if (deck.Count > 0)
            {
                //Get random card
                //Append random card to 'cards' list
                int randCard = random.Next(0, deck.Count() - 1);
                cards.Add(deck[randCard]);
                //cards.Append(deck[randCard]);
                deck.Remove(deck[randCard]);
            }
            else
            {
                deckCreation();
                int randCard = random.Next(0, deck.Count() - 1);
                cards.Add(deck[randCard]);
                //cards.Append(deck[randCard]);
                deck.Remove(deck[randCard]);
            }

            cardSum();
        }

        public void cardSum()
        {
            List<int> sumCards = new List<int>() { 0, 0 };
            cardNames.Clear();
            
            for (int c=0; c < cards.Count(); c++)
            {
                string cardName;
                int cardIdx = cards[c];
                string head = "";
                string suitName = "";
                int suitNum = ((cardIdx + 1) / 13);
                int val = (cardIdx + 1) - (suitNum * 13);

                if (suitNum <= 1)
                {
                    suitName = "Hearts";
                }
                else if(suitNum <= 2)
                {
                    suitName = "Spades";
                }
                else if(suitNum <= 3)
                {
                    suitName = "Clubs";
                }
                else
                {
                    suitName = "Diamonds";
                }

                if (val == 1)
                {
                    head = "ACE";
                }
                else if (val == 11)
                {
                    head = "KING";
                    val = 10;
                }
                else if (val == 12)
                {
                    head = "QUEEN";
                    val = 10;
                }
                else if(val == 0)
                {
                    head = "JACK";
                    val = 10;
                }

                if (head != "")
                {
                    cardName = $"{head} of {suitName}";
                }
                else
                { 
                    cardName = $"{val} of {suitName}";
                }

                cardNames.Add(cardName);
                //cardNames.Append(cardName);

                if (head == "ACE")
                {
                    sumCards[0] += 1;
                    sumCards[1] += 11;
                }
                else
                {
                    sumCards[0] += val;
                    sumCards[1] += val;
                }

            }
            handTotal = sumCards;
            busted();

        }

        public void busted()
        {
            if ((handTotal[0] == 21) || (handTotal[1] == 21))
            {
                blackJack = true;
            }
            else if ((handTotal[0] > 21) && (handTotal[1] > 21))
            {
                bust = true;
            }
        }

        public void newRound()
        {
            cards.Clear();
            cardNames.Clear();
            handTotal.Clear();
            bust = false;
            blackJack = false;
        }

        public int bestScenario()
        {
            int best = 0;
            for (int c=0; c < handTotal.Count(); c++)
            {
                if ((handTotal[c] > best) && (handTotal[c] <= 21))
                {
                    best = handTotal[c];
                }
            }
            return best;
        }
    }
}

